app/models/user.rb

```diff
@@ -13,30 +13,75 @@ class User < ApplicationRecord
-      user.full_name = access_token.extra.displayname || access_token.uid # "Harriet Tubman"
-      user.default_collection_id = Collection.default_for_department(access_token.extra.departmentnumber)&.id
+      user = new_from_cas(access_token)
+    elsif user.provider.blank?
+      user.update_with_cas(access_token)
+    end
+    user.setup_user_default_collections
+    user
+  end
+
+  # Create a new user with some basic information from CAS.
+  def self.new_from_cas(access_token)
+    user = User.new
+    user.provider = access_token.provider
+    user.uid = access_token.uid # this is the netid
+    user.email = access_token.extra.mail
+    user.display_name = access_token.extra.givenname || access_token.uid # Harriet
+    user.full_name = access_token.extra.displayname || access_token.uid # "Harriet Tubman"
+    user.default_collection_id = Collection.default_for_department(access_token.extra.departmentnumber)&.id
+    user.save!
+    user
+  end
+
+  # Updates an existing User record with some information from CAS. This is useful
+  # for records created before the user ever logged in (e.g. to gran them permissions
+  # to collections).
+  def update_with_cas(access_token)
+    self.provider = access_token.provider
+    self.email = access_token.extra.mail
+    self.display_name = access_token.extra.givenname || access_token.uid # Harriet
+    self.full_name = access_token.extra.displayname || access_token.uid # "Harriet Tubman"
+    self.default_collection_id = Collection.default_for_department(access_token.extra.departmentnumber)&.id
+    save!
+  end
+
+  # Creates a new user by uid. If the user already exists it returns the existing user.
+  def self.new_for_uid(uid)
+    user = User.find_by(uid: uid)
+    if user.nil?
+      user = User.new(uid: uid, email: "#{uid}@princeton.edu")
       user.save!
-      user.setup_user_default_collections
     end
     user
   end
 
+  # Creates the default users as indicated in the superadmin config file
+  # and the default administrators and submitters for each collection.
+  # It only creates missing records, i.e. if the records already exist it
+  # will not create a duplicate. It also does _not_ remove already configured
+  # access to other collections.
+  def self.create_default_users
+    Rails.logger.info "Setting super administrators"
+    Rails.configuration.superadmins.each { |uid| User.new_for_uid(uid) }
+
+    Collection.find_each do |collection|
+      Rails.logger.info "Setting up admins for collection #{collection.title}"
+      collection.default_admins_list.each do |uid|
+        user = User.new_for_uid(uid)
+        UserCollection.add_admin(user.id, collection.id)
+      end
+
+      Rails.logger.info "Setting up submitters for collection #{collection.title}"
+      collection.default_submitters_list.each do |uid|
+        user = User.new_for_uid(uid)
+        UserCollection.add_submitter(user.id, collection.id)
+      end
+    end
+  end
```

Comment:
When we create the default user access we create the proper User and UserCollection records in the database with the data in the `collection_defaults.yml`. These are the records that won't have a "provider" value and will be updated with the data from CAS when the user log into the system.